src/main/java/org/opensearch/security/privileges/PitAccessEvaluator.java

```diff
@@ -0,0 +1,344 @@
+                            action, resolver, presponse);
+                } else {
+                    return handleExplicitPitsAccess(pitSegmentsRequest.getPitIds(), clusterService, user, securityRoles,
+                            action, resolver, presponse);
+                }
+            }
+        } catch(InterruptedException e) {
+            Thread.currentThread().interrupt();
+            log.error(e.toString());
+        }
+        return presponse;
+    }
+
+    /**
+     * Handle access for Get All PITs access
+     */
+    private PrivilegesEvaluatorResponse handleGetAllPitsAccess(final ActionRequest request, final ClusterService clusterService,
+                                        final User user, SecurityRoles securityRoles, final String action,
+                                        IndexNameExpressionResolver resolver,
+                                        PrivilegesEvaluatorResponse presponse) throws InterruptedException {
+        List<ListPitInfo> pitInfos = getAllPitInfos((GetAllPitNodesRequest) request);
+        // if cluster has no PITs, then allow the operation to pass with empty response
+        if(pitInfos.isEmpty()) {
+            presponse.allowed = true;
+            presponse.markComplete();
+        }
+        List<String> pitIds = new ArrayList<>();
+        pitIds.addAll(pitInfos.stream().map(ListPitInfo::getPitId).collect(Collectors.toList()));
+        Map<String, String[]> pitToIndicesMap = OpenSearchSecurityPlugin.GuiceHolder.getPitService().getIndicesForPits(pitIds);
+        Map<String, ListPitInfo> pitToPitInfoMap =  new HashMap<>();
+
+        for(ListPitInfo pitInfo : pitInfos) {
+            pitToPitInfoMap.put(pitInfo.getPitId(), pitInfo);
+        }
+        List<ListPitInfo> permittedPits = new ArrayList<>();
+        for (String pitId : pitIds) {
+            String[] indices = pitToIndicesMap.get(pitId);
+            HashSet<String> indicesSet = new HashSet<>(Arrays.asList(indices));
+
+            final ImmutableSet<String> INDICES_SET = ImmutableSet.copyOf(indicesSet);
+            final IndexResolverReplacer.Resolved pitResolved =
+                    new IndexResolverReplacer.Resolved(INDICES_SET, INDICES_SET, INDICES_SET,
+                            ImmutableSet.of(), SearchRequest.DEFAULT_INDICES_OPTIONS);
+
+            final Set<String> allPermittedIndices = securityRoles.reduce(pitResolved,
+                    user, new String[]{action}, resolver, clusterService);
+            if(isDebugEnabled) {
+                log.debug("Evaluating PIT ID : " + pitId );
+            }
+            if (allPermittedIndices.size() == INDICES_SET.size()) {
+                if(isDebugEnabled) {
+                    log.debug(" Permitting PIT ID : " + pitId);
+                }
+                permittedPits.add(pitToPitInfoMap.get(pitId));
+            }
+        }
+        if (permittedPits.size() > 0) {
```

Comment:
Is there a case where `permittedPits.size()` equals to `0` as there is no PITs in the cluster, but the request should still be authorized?

Reply:
That is the case we've handled in the line 124, we authorize the user. Because the whole custom logic just ensures if PIT IDs are present, the user should have associated indices permissions.

If not, at this point the user permissions are already evaluated and user does have action permissions, so we mark operation as success.